source/extensions/priv/server/ntds_jet.c

```diff
@@ -0,0 +1,432 @@
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @param accountColumns Pointer to an ntdsState struct which will hold all of our column definitions.
+* @returns Indication of sucess or failure.
+*/
+JET_ERR next_user(struct jetState *ntdsState, struct ntdsColumns *accountColumns){
+	JET_ERR cursorStatus;
+	JET_ERR readStatus;
+	JET_ERR finalStatus = JET_errSuccess;
+	DWORD accountType = 0;
+	unsigned long columnSize = 0;
+	do{
+		cursorStatus = JetMove(ntdsState->jetSession, ntdsState->jetTable, JET_MoveNext, (JET_GRBIT)NULL);
+		if (cursorStatus != JET_errSuccess){
+			finalStatus = cursorStatus;
+			break;
+		}
+		//Retrieve the account type for this row
+		readStatus = JetRetrieveColumn(ntdsState->jetSession, ntdsState->jetTable, accountColumns->accountType.columnid, &accountType, sizeof(accountType), &columnSize, 0, NULL);
+		// Unless this is a User Account, then we skip it
+		if (readStatus == JET_wrnColumnNull){
+			continue;
+		}
+		else if (readStatus != JET_errSuccess){
+			finalStatus = readStatus;
+			break;
+		}
+	} while (accountType != 0x30000000);
+	return finalStatus;
+}
+
+/*!
+* @brief Attach our Jet Instance to the ntds.dit file and open the 'datatable' table for reading.
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @returns Indication of sucess or failure.
+*/
+JET_ERR open_database(struct jetState *ntdsState){
+	JET_ERR attachStatus = JetAttachDatabase(ntdsState->jetSession, ntdsState->ntdsPath, JET_bitDbReadOnly);
+	if (attachStatus != JET_errSuccess){
+		return attachStatus;
+	}
+	JET_ERR openStatus = JetOpenDatabase(ntdsState->jetSession, ntdsState->ntdsPath, NULL, &ntdsState->jetDatabase, JET_bitDbReadOnly);
+	if (openStatus != JET_errSuccess){
+		return openStatus;
+	}
+	return JET_errSuccess;
+}
+
+/*!
+* @brief Read the current user record into an ntdsAccount struct.
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @param accountColumns Pointer to an ntdsState struct which will hold all of our column definitions.
+* @param pekDecrypted Pointer to a decryptedPEK structure that holds our decrypted PEK
+* @param userAccount Pointer to an ntdsAccount struct that will hold all of our User data
+* @returns Indication of sucess or failure.
+*/
+JET_ERR read_user(struct jetState *ntdsState, struct ntdsColumns *accountColumns, struct decryptedPEK *pekDecrypted, struct ntdsAccount *userAccount){
+	JET_ERR readStatus = JET_errSuccess;
+	DWORD accountControl = 0;
+	unsigned long columnSize = 0;
+	// Grab the SID here
+	readStatus = JetRetrieveColumn(ntdsState->jetSession, ntdsState->jetTable, accountColumns->accountSID.columnid, &userAccount->accountSID, sizeof(userAccount->accountName), &columnSize, 0, NULL);
+	if (readStatus != JET_errSuccess){
+		return readStatus;
+	}
+	// Derive the RID from the SID
+	int ridIndex = columnSize - sizeof(DWORD);

```

Comment:
If the above JetRetrieveColumn could return 0 (I don't know if can, just warning) I'd say you're in trouble here